src/extension_traits/vec.rs

```diff
@@ -0,0 +1,244 @@
+use crate::*;
+use ::core::slice;
+
+/// Extension trait for [`Vec`], allowing a non-`unsafe` API to interact
+/// with the backing buffer / allocation.
+pub
+trait VecAllocation : Sealed {
+    type Item;
+    fn get_backing_buffer (self: &'_ mut Self)
+      -> OutSlice<'_, Self::Item>
+    ;
+    fn into_backing_buffer (self: Self)
+      -> Vec<MaybeUninit<Self::Item>> // ideally this would be a `Box<[]>`, but `Vec` does not guarantee it.
+    ;
+    fn split_at_extra_capacity (self: &'_ mut Self)
+      -> (&'_ mut [Self::Item], &'_ mut [MaybeUninit<Self::Item>])
+    ;
+    fn reserve_uninit (self: &'_ mut Self, additional: usize)
+      -> &'_ mut [MaybeUninit<Self::Item>]
+    ;
+}
+
+mod private { pub trait Sealed : Sized {} } use private::Sealed;
+
+impl<T> Sealed for Vec<T> {}
+impl<T> VecAllocation for Vec<T>
+where
+    T : Copy, // Opinionated position against accidental memory leaks
```

Comment:
TODO: Add helpers for non-`Copy` types (_e.g._, `&mut Vec<T> -> &mut Vec<ManuallyDrop<T>>` conversion and `impl` the trait for `Vec<ManuallyDrop<T>>`